benchmark/src/util.ts

```diff
@@ -0,0 +1,85 @@
+import { BenchmarkFunction } from "./benchmark_types";
+
+export function round(num: number, decimalPlaces: number = 0) {
+    return tonumber(string.format(`%.${decimalPlaces}f`, num));
+}
+
+export const json: {
+    decode: (this: void, str: string) => {};
+    encode: (this: void, val: any) => string;
+} = require("json");
+
+export function readFile(path: string): string | undefined {
+    const fileOpenArray = io.open(path, "rb");
+
+    if (fileOpenArray && fileOpenArray[0]) {
+        const fileHandle = fileOpenArray[0];
+        let fileContent = readAll(fileHandle);
+        fileHandle.close();
+
+        if (fileContent) {
+            return fileContent;
+        }
+    }
+}
+
+export function readAll(file: LuaFile): string | undefined {
+    let content: (string | undefined)[];
+    if (_VERSION == "Lua 5.3") {
+        // @ts-ignore
+        content = file.read("a");
+    } else {
+        // JIT
+        // @ts-ignore
+        content = file.read("*a");
+    }
+
+    if (content && content[0]) {
+        return content[0];
+    }
+}
+
+export function readDir(dir = ""): string[] | undefined {
+    let isWindows = false;
+    let [success, findHandle] = pcall(() => io.popen(`find '${dir}' -maxdepth 1 -type f | sed "s|^\./||"`));
+
+    if (!success) {
+        [success, findHandle] = pcall(() => io.popen(`dir /A-D /B ${dir}`));
+        isWindows = true;
+    }
+
+    if (success) {
+        // appereantly TS can't infer this on it's own
```

Comment:
I doubt TS would ever start tracking dependencies between unrelated variables. Instead of casting we might change code a bit to make it inferrable https://typescripttolua.github.io/play#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVysAQwBMARTKACiPLgF44AiBgSgC44BnGKTFAcwDaAXTgAfOKiLAEvYETgBvAFBxVcADbB4mDgHVeRCAHcOdRAXUdgAbhVrN8SGGAoASsA5J18emDQX1CgpmOgA+OGwAOicXCgADGRR5AHIAEgVqKABfZLgAWgBbAhApMBgACzgARnyYAE9nRDFOOUYOUQA9AB1IgHpRUQY45mZbO1VMBDgKAEIYtw8vGAEABiEQ5TUtuHn3T28zPwCgkNpwqPn4zLhegEE8khuAITh0zKzh0fHtnX0k41M9G4SBs3yySm+k2mu0W3lW60U3y2vV6cAIYGc+AIKBg6jqcAAKgBlOD+FDJbQoBDAKBwCo6OA4CIwZKmYwoJFqNA4LiIAwACWxRE0hwgzgW+2WVSEY22XJ58ESRD2SzMWKIt3UgSVgqSmi+ctUOqFmkiaHUECswVlhqhFCVKu8G05cocfM0gL5SUdMEiHDA6kwMAoDC6KBYtkNPymFF+BgBzqjUZRjNwRnjJjgRmAcBQwFaMAgaIxLnkFRzmWAGGg+LABAqLqTKaZqEwIDp5QZDMMebgACN8Uq0UhC0UsP4tXVG1GZB6zLOPJEimB7WE4HE3uQsr10ggPgak6pwYfVPgYEgoLgHbDff7A8HQ+HmJFZzAaRQwGvPzNaPQmAe5WPNRwSyIA